[LuxtronikHeatpump] Adds new binding for integrating heat pumps based on a Luxtronik control unit

sgiehl opened the pull request:
_First of all - I'm kind of new to Java, or at least didn't do anything with it in the last 20 years. Started using OpenHab a few months ago and started looking deeper into implementing code changes 3 weeks ago. So please bear with me if some things are not implemented optimally. If anything should be changed or can be improved please leave some comments and I will try to adjust it._

The initial intention of building this was to migrate the [novelanheatpump](https://github.com/openhab/openhab1-addons/tree/master/bundles/binding/org.openhab.binding.novelanheatpump) binding for OpenHab3. While trying to migrate the code I realised that there was a lot space for improvements. I've decided to give the binding a new name, as it is actually not only able to handle Novelan heatpumps.

Based on some research on the net the binding should work with all heat pumps having a Luxtronik control unit. As already mentioned in the README, that should be:

* Alpha InnoTec
* Buderus (Logamatic HMC20, HMC20 Z)
* CTA All-In-One (Aeroplus)
* Elco
* Nibe (AP-AW10)
* Roth (ThermoAura®, ThermoTerra)
* (Siemens) Novelan (WPR NET)
* Wolf Heiztechnik (BWL/BWS)

I'm developing / testing it myself with a Siemens Novelan LD 7.

A built version can be found here: http://stefangiehl.de/openhab/org.openhab.binding.luxtronikheatpump-3.1.0-SNAPSHOT.jar
I'll try to update it with each commit I push.

Compared to the `novelanheatpump` binding this binding already has this additionally:

* **Additional channels**
  The binding currenlty has channels for almost all (read-only) values returned by the heatpump socket. It also implements some additional writable channels

* **Channel visibilities**
  The heat pump socket is able to serve a lot values that define visibilities. Those are used by the internal java applet to determine which settings to show. I tried to add a check for almost all channels based on those visibility settings, so channels are automatically hidden when they are not available. As it's hard to know if all checks are correct, I've added an additional thing config `showAllChannels`. When enabled, all channels will be shown and visibilities will be ignored.

The current state of this PR should be already working correctly.
Nevertheless I've got this tasks on my TODO, some of them possibly for follow-up PRs:

* **Implement more writable channels**
   The amount of writable channels is currently limited to those that were available in the `novelanheatpump` binding. But actually my heat pump is able to receive over 1000 possible parameters. A full list can be found [here](https://www.loxwiki.eu/download/attachments/13306044/3003.txt?version=1&modificationDate=1487705234000&api=v2). Will postpone the research on this for now, as it wouldn't be needed for the initial PR though.

* **Adjust / Complete README**
  The README is not yet finished. It lags some examples

* **Discovery**
  Discovery of heat pumps in the network could maybe be possible. Did not yet have time to check how to implement that properly. But might be something for a follow-up PR.

* **Testing**
  ~I need to find some time to validate the values returned by the binding. Need to look through the settings and values and compare them to those displayed by the heat pump itself (or the Java applet that's shipped on heat pump's internal webserver)~ done. Tested at least those values/settings that are available on my heatpump


Discussion on the forum can be found here: https://community.openhab.org/t/migration-of-novelan-luxtronic-binding-to-openhab2/70743

Hilbrand commented:
I believe the brandingname is `luxtronik`

sgiehl commented:
@Hilbrand your're right. will change the name

sgiehl commented:
Can anyone explain me why the build is failing?

Hilbrand commented:
It looks like it fails on the copyright year. This is because the build first merges and that will change the copyright year check to 2021. You need to rebase on the latest code, run `mvn spotless:apply` on your binding and push force back.

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpVisibility.java:
```diff
@@ -0,0 +1,378 @@
+            }
+        }
+
+        throw new IllegalArgumentException("Invalid heat pump visibility code: '" + code + "'");
```
as above

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpVisibility.java:
```diff
@@ -0,0 +1,378 @@
+    SPRACHEN_LETTISCH(283, "Sprachen_LETTISCH"),
+    SPRACHEN_FINNISCH(284, "Sprachen_FINNISCH"),
+    KALIBRIERUNG_LWD(285, "Kalibrierung_LWD"),
+    IN_DURCHFLUSS(286, "IN_Durchfluss"),
+    LIN_ANSAUG_VERDICHTER(287, "LIN_ANSAUG_VERDICHTER"),
+    LIN_VDH(288, "LIN_VDH"),
+    LIN_UH(289, "LIN_UH"),
+    LIN_DRUCK(290, "LIN_Druck"),
+    EINST_SOLLWERT_TRL_KUEHLEN(291, "Einst_Sollwert_TRL_Kuehlen"),
+    ENTL_EXVENTIL(292, "Entl_ExVentil"),
+    EINST_MEDIUM_WAERMEQUELLE(293, "Einst_Medium_Waermequelle"),
+    EINST_MULTISPEICHER(294, "Einst_Multispeicher"),
+    EINST_MINIMALE_RUECKLAUFSOLLTEMPERATUR(295, "Einst_Minimale_Ruecklaufsolltemperatur"),
+    EINST_PKUEHLTIME(296, "Einst_PKuehlTime"),
+    SPRACHEN_TUERKISCH(297, "Sprachen_TUERKISCH"),
+    RBE(298, "RBE"),
+    EINST_LUF_STUFEN_FAKTOR(299, "Einst_Luf_Stufen_Faktor"),
+    FREIGABE_ZEIT_ZWE(300, "Freigabe_Zeit_ZWE"),
+    EINST_MIN_VL_KUEHL(301, "Einst_min_VL_Kuehl"),
+    ZWE1(302, "ZWE1"),
+    ZWE2(303, "ZWE2"),
+    ZWE3(304, "ZWE3"),
+    SEC(305, "SEC"),
+    HZIO(306, "HZIO"),
+    WPIO(307, "WPIO"),
+    LIN_ANSAUG_VERDAMPFER(308, "LIN_ANSAUG_VERDAMPFER"),
+    LIN_MULTI1(309, "LIN_MULTI1"),
+    LIN_MULTI2(310, "LIN_MULTI2"),
+    EINST_LEISTUNG_ZWE(311, "Einst_Leistung_ZWE"),
+    SPRACHEN_ESPANOL(312, "Sprachen_ESPANOL"),
+    TEMP_BW_OBEN(313, "Temp_BW_oben"),
+    MAXIO(314, "MAXIO"),
+    OUT_ABTAUWUNSCH(315, "OUT_Abtauwunsch"),
+    SMARTGRID(316, "SmartGrid"),
+    DREHZAHLGEREGELT(317, "Drehzahlgeregelt"),
+    P155_INVERTER(318, "P155_Inverter"),
+    LEISTUNGSFREIGABE(319, "Leistungsfreigabe"),
+    EINST_VORL_AKT_KUEHL(320, "Einst_Vorl_akt_Kuehl"),
+    EINST_ABTAUEN_IM_WARMWASSER(321, "Einst_Abtauen_im_Warmwasser"),
+    WAERMEMENGE_ZWE(32, "Waermemenge_ZWE");
```
language as above.

sgiehl replied:
The enum names are those used in the internal java applet of the heat pump. Don't think it makes much sense to translate them.

daMihe replied:
The question int this case is: Is this always the case? Is the language switchable in the device or can other people get different names? If yes, the more internationally used language should win. If the channels are even on outside of DACH sold devices are named in german, i it is ok to name them in german also here.

sgiehl replied:
Guess I didn't make that clear. Those are the variable names used inside the java applet code. Those aren't visible to the customer at all. The java applet itself is served with translation files. But as far as I have seen for some languages not all words are translated (even not for English). Some are still German and some translations are built by concatenating various parts, which imho might not be correct for all languages. But it's hard to say. The code of the applet is a jar file containing compiled classes...

daMihe replied:
Do i understand right? Those enum constants are the reverse engineered internal variable names? Then, i guess, the german strings would be okay. But please document them then using Javadoc-comments to describe them in English.

I'm not fully sure if this is enough. There we need to wait on a second review of a more experienced reviewer. But i'm pretty sure just german names won't be enough.

sgiehl replied:
For those visibility values I actually don't have any proper descriptions. Not even in german. I'm guessing their meaning based on the return values of my heat pump and the internal names. But will add a comment to describe that.

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,388 @@
+channel-type.luxtronikheatpump.ERROR_NrX.state.option.790 = Schwerw. Inverter Fehler
+channel-type.luxtronikheatpump.ERROR_NrX.state.option.791 = ModBus Verbindung verloren
+channel-type.luxtronikheatpump.ERROR_NrX.state.option.792 = LIN-Verbindung unterbrochen
+channel-type.luxtronikheatpump.ERROR_NrX.state.option.793 = Schwerw. Inverter Fehler
```
Please add newline at end

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump.properties:
```diff
@@ -0,0 +1,84 @@
+channel-type.luxtronikheatpump.ERROR_NrX.state.option.790 = Major VSS fault
+channel-type.luxtronikheatpump.ERROR_NrX.state.option.791 = Lost ModBus communication
+channel-type.luxtronikheatpump.ERROR_NrX.state.option.792 = LIN-connection lost
+channel-type.luxtronikheatpump.ERROR_NrX.state.option.793 = Major VSS fault
```
Please add newline at end

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java:
```diff
@@ -0,0 +1,784 @@
+            }
+        }
+
+        throw new IllegalArgumentException("cannot find LuxtronikHeatpump channel for '" + heatpumpCommand + "'");
```
Unchecked exception

daMihe replied:
Unchecked does not mean it's missing in throws section of method signature. Instead it does mean the compiler does not force you to write a try block around it, but that would be better. Instead of reusing `InvalidArgumentExcpetion` use a different exception type or define you own. (also for the other places.)

sgiehl replied:
ok. will replace that with a custom exception

daMihe reviewed (changes requested):
This is my first review. So  there are possibly many other things to fix.

Please note that  openHABs main language is english. So channels and variables should be also implemented in english. You are are a bit mixing german and english. This makes it hard for other possible contributors can add here something.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/README.md:
```diff
@@ -0,0 +1,278 @@
+* (Siemens) Novelan (WPR NET)
+* Wolf Heiztechnik (BWL/BWS)
+
+This binding is based on the [Novelan/Luxtronik Heat Pump Binding](https://v2.openhab.org/addons/bindings/luxtronikheatpump1/) for Open Hab 1. And some other Luxtronik tools like [Luxtronik2 for NodeJS](https://github.com/coolchip/luxtronik2) and the detailed parameter descriptions for the Java Webinterface in the [Loxwiki](https://www.loxwiki.eu/display/LOX/Java+Webinterface)
```
for me the link is still working :thinking:

daMihe replied:
I'm getting a 404 when visiting the first link. As the Binding is no more usable in openHAB 3 i'm not sure its confusing there. If you want to attribute others works you could add a section at the end of the Readme just listing what taken or inspired from where.

sgiehl replied:
will move the part to the end and fix the link

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/README.md:
```diff
@@ -0,0 +1,278 @@
+
+## Channels
+
+As the Luxtronik 2 control is able to handle multiple heat pumps with different features (like heating, hot water, cooling, solar, photovoltaics, swimming pool,...), the binding has a lot channels. Depending on the heatpump it is used with, various channels might not hold any (useful) values. If `hideChannels` is activated for the thing, this binding will automatically try to hide channels that a ...
```
Thought it would make more sense to have only usable channels available. Having multiple things might be another option, but guess it would be very hard to logically split it into things and split the channels between those things.

daMihe replied:
Okay, please use the word "generic" then in types and ids. This allows later to add specialised devices for less confusion.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/ChannelUpdaterJob.java:
```diff
@@ -0,0 +1,211 @@
+                    continue; // no channel id to read defined (for channels handeled separatly)
+                }
+
+                if (channel.isVisible(heatpumpVisibilities).equals(Boolean.FALSE) && config.hideChannels) {
```
Is it enough to change the default value in `LuxtronikHeatpumpConfiguration` class to `true`?

daMihe replied:
I'm not sure in this case. Please try.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/ChannelUpdaterJob.java:
```diff
@@ -0,0 +1,211 @@
+
+            updateProperties(heatpumpValues);
+
+        } catch (Exception e) {
+            logger.warn("Could not connect to heatpump (uuid={}, ip={}, port={}): {}", thing.getUID(), config.ipAddress,
+                    config.port, e.getStackTrace());
+        }
```
Seems I forgot to change that back again. Had added the stackTrace instead of the message for debugging purpose only.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/ChannelUpdaterJob.java:
```diff
@@ -0,0 +1,211 @@
+        }
+
+        if (itemClass == NumberItem.class) {
+            if (unit == SIUnits.CELSIUS) {
+                return new QuantityType<>((double) rawValue / 10, SIUnits.CELSIUS);
+            } else if (unit == Units.KELVIN) {
+                return new QuantityType<>((double) rawValue / 10, Units.KELVIN);
+            } else if (unit == Units.KILOWATT_HOUR) {
+                return new QuantityType<>((double) rawValue / 10, Units.KILOWATT_HOUR);
+            } else if (unit == Units.PERCENT) {
+                return new QuantityType<>((double) rawValue / 10, Units.PERCENT);
+            } else if (unit == Units.HOUR) {
+                return new QuantityType<>((double) rawValue / 10, Units.HOUR);
+            } else if (unit == Units.HERTZ) {
+                return new QuantityType<>((double) rawValue, Units.HERTZ);
+            } else if (unit == Units.SECOND) {
+                return new QuantityType<>((double) rawValue, Units.SECOND);
+            } else if (unit == Units.LITRE_PER_MINUTE) {
+                return new QuantityType<>((double) rawValue * 60, Units.LITRE_PER_MINUTE);
+            } else if (unit == Units.BAR) {
+                return new QuantityType<>((double) rawValue / 100, Units.BAR);
+            } else if (unit == Units.VOLT) {
+                return new QuantityType<>((double) rawValue / 100, Units.VOLT);
+            } else {
+                return new DecimalType(rawValue);
+            }
```
actually tried that before, but vscode warns `Cannot switch on a value of type Unit<capture#4-of ?>. Only convertible int values, strings or enum variables are permitted` and mvn fails with errors...

daMihe replied:
Okay, maybe because of the different types of Units. You could at least merge the same handled units to greater or-ed conditions.

sgiehl replied:
ok. will try to combine them a bit

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,280 @@
+        super(thing);
+    }
+
+    @Override
+    public void updateState(String channelID, State state) {
+        super.updateState(channelID, state);
+    }
+
+    @Override
+    public void updateProperty(String name, String value) {
+        super.updateProperty(name, value);
+    }
```
they are protected in the super class. As updating is done in another class I made them public to have them available.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,280 @@
+                }
+            }
+            scheduledFutures.clear();
+        } catch (Exception ex) {
+            logger.error("{}", ex.getMessage(), ex);
```
That part is actually copied from the Astro binding :see_no_evil: Not sure if the code inside the try could throw an exception at all :shrug:

sgiehl commented:
@daMihe thanks for the review. I've already applied some changes. Will check the remaining ones tomorrow.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/README.md:
```diff
@@ -0,0 +1,278 @@
+| Einst_KuCft1_akt | Number:Temperature |   | Comfort cooling AT release | 
+| Sollwert_KuCft1_akt | Number:Temperature |   | Comfort cooling AT release target | 
+| Einst_Kuhl_Zeit_Ein_akt | Number |   | AT Excess | 
+| Einst_Kuhl_Zeit_Aus_akt | Number |   | AT undercut | 
```
The channel names are currently those that are used by the internal java applet of the heat pump. That makes it easier to match them with the visibility values. If needed we could change at least the visible names to english

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandlerFactory.java:
```diff
@@ -0,0 +1,67 @@
+public class LuxtronikHeatpumpHandlerFactory extends BaseThingHandlerFactory {
+
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_HEATPUMP);
+    private static final Map<String, LuxtronikHeatpumpHandler> THING_HANDLERS = new HashMap<>();
```
No. It actually maps the things to the handler instances. That needs to be static

daMihe replied:
Sorry, this is wrong knowing a bit the framework and how static variables are initialised and destroyed. `getHandler(String)` is not used by the framework so it is not required there. Static variables do not have a defined lifetime. This is bad as it could lead to side effects and cause other trouble (de facto you keep instances beside of the framework alive which is very bad). Please use the frameworks possibilities for these purposes.

You are using your `getHandler(String)` only in `ChannelUpdaterJob`. There you don't even need to access the `LuxtronikHeatpumpHandlerFactory` as you already have access to `thing`. `Thing` (which `thing` is an instance of) has a non-static method `getHandler()` which just links to the proper instance. Using it this way (and properly removing the schedule when disposing the Handler) keeps the Lifecycle working as defined by openHAB.

sgiehl replied:
changed that

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,280 @@
+public class LuxtronikHeatpumpHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(LuxtronikHeatpumpHandler.class);
+    private final Lock monitor = new ReentrantLock();
```
I'll try to change that

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,280 @@
+        // connect to heatpump and check if values can be fetched
+        HeatpumpConnector connector = new HeatpumpConnector(config.ipAddress, config.port);
+
+        try {
+            connector.read();
+        } catch (UnknownHostException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "the given hostname ''" + config.ipAddress + "' of the heatpump is unknown");
+            return;
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "couldn't establish network connection [host '" + config.ipAddress + "']");
+            return;
+        }
```
Tried to improve the behaviour so the initialize will be retried until it succeeds.

daMihe replied:
This is a bit better, but not optimal. Networking is unreliable. You have to deal with packet loss (TCP does for you), disconnects (OS just notifies the application), unexpected latency (which should be uncritical here) etc.. The better way would be to start or restart (if necessary) the connection when communicating with the device. You could put the connection initialisation into a method which also gets called whenever data is read and check whether there is a socket and it's connected.

daMihe replied:
I've seen the connection is already built when reading. That is good. But the state of the thing should correspond to the current (or last tried last) state. Also retrying every 0 seconds is not optimal. You already have defined an interval for updating. You could use it here or combine the update of channels with the thing state.

sgiehl replied:
I'll add updating the status in the updater job as well. The initial state detection should be done without a delay but retried every 60 seconds if it isnn't cleared.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,280 @@
+            }
+        }
+
+        updateThing(thingBuilder.build());
```
Tried to change that. But actually I realized that there is another problem with that logic. Changing the setting to show all channels after the thing was already initialized doesn't work correctly. Guess the channels are already removed from that thing and aren't added back automatically. Did not yet find a way to reload the channels from the xml definitions. Not sure if there is any :thinking:

daMihe replied:
You can add them again using `ThingBuilder.withChannel()`, using the right type, the framework reads out the XML descriptions. I've done this in icalendar [there](https://github.com/openhab/openhab-addons/blob/0816045a5949de092c7a4ea9e25467fffdde6b6b/bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/ICalendarHandler.java#L281-L285) with a single channel. The preferred method would be to prepare the list of channels, that need to be added again and add them in a batch to ensure the thing-edit is done in a short time. I guess the time should be short as there is no lock or so.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java:
```diff
@@ -0,0 +1,784 @@
+    // for possible values see https://www.loxwiki.eu/display/LOX/Java+Webinterface
+    // or https://github.com/Bouni/Home-Assistant-Luxtronik/blob/master/data.txt
+
+    // Vorlauftemperatur Heizkreis
```
Added the english translations there as well now

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java:
```diff
@@ -0,0 +1,968 @@
+    // for possible values see https://www.loxwiki.eu/display/LOX/Java+Webinterface
+    // or https://github.com/Bouni/Home-Assistant-Luxtronik/blob/master/data.txt
+
+    // EN: Flow temperature heating circuit
+    // DE: Vorlauftemperatur Heizkreis
```
```suggestion
    /**
     * Flow temperature heating circuit (original: Vorlauftemperatur Heizkreis)
     */
```

Please format these descriptions as Javadoc. This way the IDE's documentation framework can work.

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/HeatpumpConnector.java:
```diff
@@ -0,0 +1,195 @@
+     * @throws IOException indicate that no data can be read from the heat pump
+     */
+    public void read() throws UnknownHostException, IOException {
+        try (Socket sock = new Socket(serverIp, serverPort)) {
```
I'm not sure whether the protocol allows it to keep the connection alive. If yes, you could keep the connection open and reduce the risk of later issues when DNS or other network things are failing but the communication itself would work.

sgiehl replied:
I'm not able to answer that. The old binding and the other tools, I found for managing a luxtronik heatpump, are closing the connection again. In order to avoid any possible problems if the socket does not support that properly I wouldn't change that for now.

Hilbrand commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,340 @@
+        String channelId = channelUID.getIdWithoutGroup();
+        logger.debug("Handle command '{}' for channel {}", command, channelId);
+        if (command == RefreshType.REFRESH) {
+            restartJobs();
```
`REFRESH` is called for every channel. So if you have 20 channels you will restart the job 20 times. Refresh should be used to call updateState on the given channel with either a cached value or get the value from the device (although that might be expensive) or just ignore it, if it's handled by the `channelUpdaterJob`.

sgiehl replied:
Ah ok. Wasn't aware of that. Guess removing it is the best choice then. Will do that later.

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/ChannelUpdaterJob.java:
```diff
@@ -0,0 +1,285 @@
+    @Override
+    public void run() {
+        // connect to heatpump and check if values can be fetched
+        final HeatpumpConnector connector = new HeatpumpConnector(config.ipAddress, config.port);
+        final LuxtronikHeatpumpHandler handler = (LuxtronikHeatpumpHandler) thing.getHandler();
```
This could be moved into constructor including the cast and the non null check so you don't need to initialise the members every time and can assume they are non-null.

sgiehl replied:
will change it to a constructor parameter

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/ChannelUpdaterJob.java:
```diff
@@ -0,0 +1,285 @@
+                    handleEventType(value, channel);
+                }
+
+            } catch (Exception e) {
+                logger.warn("An error occurred while updating the channel {}: {}", channel.getCommand(),
+                        e.getMessage());
+            }
```
can you be more specific here?

sgiehl replied:
actually there should no exception be thrown. At least it didn't while testing. It's catched nevertheless so following channels will be updated even if one fails. Will add that as a comment in the code so it clear.

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/HeatpumpConnector.java:
```diff
@@ -0,0 +1,195 @@
+            datain.readInt();
+        }
+
+        int arraylength = datain.readInt();
```
Just a thought: Maybe an upper limit would be useful as int's max could lead to try to allocate ~8G RAM (assuming an `Integer`-Pointer is 4 Bytes because of an 32-Bit system)

sgiehl replied:
You mean in case the socket returns (for any reason) a number that is a lot higher than expected. Shouldn't happen I think. But won't hurt to add a check

daMihe reviewed (commented):
Some further comments while re-reviewing.

fwolter commented:
@daMihe Are all of your comments addressed?

daMihe commented:
> @daMihe Are all of your comments addressed?

I'll look tomorrow again.

daMihe commented:
@fwolter With exception of README it looks good to me (at least github shows no changes in a few commented sections). Please note that is my first review.

sgiehl commented:
@daMihe I've now added an example to the readme.

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/README.md:
```diff
@@ -0,0 +1,328 @@
+        Text item=HeatPump_Hours_Heatpump
+        Text item=HeatPump_Hours_Heating
+        Text item=HeatPump_Hours_Warmwater
+        Switch item=HeatPump_heating_operation_mode  mappings=[0="Auto", 1="Auxiliary heater", 2="Party", 3="Holliday", 4="Off"]
```
Just a minimal typo: "Holiday"

daMihe replied:
```suggestion
        Switch item=HeatPump_heating_operation_mode  mappings=[0="Auto", 1="Auxiliary heater", 2="Party", 3="Holiday", 4="Off"]
```

daMihe commented on bundles/org.openhab.binding.luxtronikheatpump/README.md:
```diff
@@ -0,0 +1,328 @@
+        Text item=HeatPump_Hours_Warmwater
+        Switch item=HeatPump_heating_operation_mode  mappings=[0="Auto", 1="Auxiliary heater", 2="Party", 3="Holliday", 4="Off"]
+        Setpoint item=HeatPump_heating_temperature minValue=-10 maxValue=10 step=0.5
+        Switch item=HeatPump_warmwater_operation_mode  mappings=[0="Auto", 1="Auxiliary heater", 2="Party", 3="Holliday", 4="Off"]
```
```suggestion
        Switch item=HeatPump_warmwater_operation_mode  mappings=[0="Auto", 1="Auxiliary heater", 2="Party", 3="Holiday", 4="Off"]
```

remogloor commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpType.java:
```diff
@@ -0,0 +1,128 @@
+    TYPE_LWD90V(77, "LWD90V"),
+    TYPE_MSW3_12(78, "MSW3-12"),
+    TYPE_MSW3_12S(79, "MSW3-12S"),
+    TYPE_UNKNOWN(-1, "Unknown");
```
Heatpump Type 80 is missing
80 = MSW2-9S

sgiehl replied:
@remogloor Do you have that heat pump? Just wondering where you got that information from and if there might be some more types missing.

remogloor replied:
@sgiehl Yes I have a Alpha Innotec SWCV 92H3 which has the name MSW2-9S in Luxtronic. The binding complains in the logs that type 80 is unknown. Therefore I assume that type 80 should be MSW2-9S.

remogloor replied:
@sighl By the way, the unnecessary logs is the only problem of the missing type definition. Except that this is a bit annoying as it spams the logs everything else works perfectly. I could not yet see another problem caused by the missing definition.

sgiehl replied:
@remogloor Thanks for letting me know. Guess my heatpump has an older software version maybe. Could you check if your heatpump also serves a java applet? If possible it would be helpful if you could send me the applet by mail (stefan@matomo.org). You can download the applet at `http://{yourheatpumpip}/wp.jar` (at least that is the location on my heatpump)
Maybe the code includes some more missing definitions.

sgiehl replied:
@remogloor I've added the additional type and updated the file on my server. Feel free to send me the jar file nevertheless if possible.

pbolle commented:
Hello, I have probably done the work for nothing. I started to port the novelanheatpump binding (https://github.com/openhab/openhab-addons/pull/10178). But you were probably faster. Then I wait and see what happens here.

sgiehl commented:
@pbolle Maybe you could do a quick code review on this one and/or maybe test the binding if it works for you.

sgiehl commented:
@fwolter @Hilbrand @wborn Are there any plans on further reviewing the PR so it gets merged maybe before the next release?
Based on the comments here and on the forum it's seems to work as excepted so far. Let me know if anything needs to be adjusted. If any issues would come up in the future I'd be happy to fix them if I can...

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,344 @@
+
+    private final Logger logger = LoggerFactory.getLogger(LuxtronikHeatpumpHandler.class);
+    private final Set<ScheduledFuture<?>> scheduledFutures = new HashSet<>();
+    private static final int RETRY_INTERVAL = 60;
```
It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    public void setStatusConnectionError() {
+        LuxtronikHeatpumpConfiguration config = getConfigAs(LuxtronikHeatpumpConfiguration.class);
```
As this is an expensive operation, you could store it once in `initialize()` in a field.

sgiehl replied:
moved it to the constructor. guess that should work as well.

fwolter replied:
When the user changes the configuration during runtime, only `initialize()` is invoked, but not the constructor.

sgiehl replied:
makes sense. Would it be fine to load the config in constructor and overwrite it in initialize again? Otherwise I guess I would need to mark the field as nullable and check if it's set everytime it is used, or is there a smarter solution?

fwolter replied:
You can simply initialize the field with an empty configuration to avoid the null checks.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,344 @@
+        }
+
+        if (command instanceof QuantityType) {
+            command = new DecimalType(((QuantityType<?>) command).floatValue());
```
Did you consider that the return value of `floatValue()` will be Fahrenheit if the user uses Fahrenheit? You can convert it to a unit you expect by calling `((QuantityType<?>) command).toUnit()`.

sgiehl replied:
good point. Didn't think of that. will change it

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,344 @@
+        updateStatus(ThingStatus.UNKNOWN);
+
+        synchronized (scheduledFutures) {
+            logger.debug("try to connect to the heat pump every {} seconds to update the status", RETRY_INTERVAL);
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

sgiehl replied:
I'm not sure what you are trying to point out. I had a look at binding like the one for Fritz!Box while developing. It also uses logger.debug to log some stuff that is happening, like it is in this case. Shall that be removed or is there only another way it should be done?

fwolter replied:
It's uncommon in bindings to log a periodic task has been started. You could remove it.

sgiehl replied:
ok. will remove it

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,344 @@
+        if (thing.getProperties().isEmpty() || tiggerChannelUpdate) {
+            try {
+                updateChannels(connector);
+            } catch (Exception e) {
```
What type of exception do you expect here? Can you specify the concrete type?

sgiehl replied:
It's been a while since I've added that. Guess it was to make sure initalizing the thing can't fail at that point and the job to update the channels is set up even if it fails to update the channels here for some reason.

fwolter replied:
It seems the method doesn't throw any reasonable exception. If it throws an Error, there might be something wrong fundamentally. This case is handled by the framework by logging all unchecked und uncought exceptions.

sgiehl replied:
let's assume a network timeout happens here as the device or network is temporary unavailable. Not catching the exception would result in not setting up the periodic update task. Shouldn't the exception been caught and instead the task been set up using at `try { } finally {}` instead?

fwolter replied:
If I see correctly, `updateChannels()` doesn't perform any network actions. It also doesn't declare any `throws` in its method signature.

sgiehl replied:
ah. you're right. It used to trigger some network actions, but I removed that as part of another refactoring. Will remove the try catch

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,344 @@
+
+    @Override
+    public void dispose() {
+        logger.debug("Disposing thing {}", getThing().getUID());
+        stopJobs();
+        logger.debug("Thing {} disposed", getThing().getUID());
```
As this is already logged by the framework, the log messages could be removed.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    private void updateChannels(HeatpumpConnector connector) {
+        LuxtronikHeatpumpConfiguration config = getConfigAs(LuxtronikHeatpumpConfiguration.class);
```
See above. Make it a field.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,344 @@
+        // create list with available channels
+        for (HeatpumpChannel channel : HeatpumpChannel.values()) {
+            Integer channelId = channel.getChannelId();
+            int length = channel.isWritable() == Boolean.TRUE ? heatpumpParams.length : heatpumpValues.length;
```
Same for below.
```suggestion
            int length = channel.isWritable() ? heatpumpParams.length : heatpumpValues.length;
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java:
```diff
@@ -0,0 +1,344 @@
+                ScheduledFuture<?> future = scheduler.scheduleWithFixedDelay(channelUpdaterJob, 0, config.refresh,
+                        TimeUnit.SECONDS);
+                scheduledFutures.add(future);
+                logger.debug("Scheduled {} every {} seconds", channelUpdaterJob, config.refresh);
```
See above. Replace logging by debugger. Same for below.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandlerFactory.java:
```diff
@@ -0,0 +1,69 @@
+    @Activate
+    public LuxtronikHeatpumpHandlerFactory(final @Reference LocaleProvider localeProvider,
+            final @Reference TranslationProvider i18nProvider, ComponentContext componentContext) {
+        super.activate(componentContext);
```
Are you sure this call is necessary?

sgiehl replied:
guess that was a copy/paste from some example. will check if it can be removed...

sgiehl replied:
seems removing it results in this error when the binding is loaded:
```
2021-03-21 17:53:27.239 [ERROR] [rg.openhab.binding.luxtronikheatpump] - bundle org.openhab.binding.luxtronikheatpump:3.1.0.202103211652 (209)[org.openhab.binding.luxtronikheatpump.internal.LuxtronikHeatpumpHandlerFactory(361)] :  Error during instantiation of the implementation object
java.lang.reflect.InvocationTargetException: null
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[?:?]
	at org.apache.felix.scr.impl.inject.ComponentConstructor.newInstance(ComponentConstructor.java:309) ~[bundleFile:?]
...

```

fwolter replied:
That's strange. This code works for me:
```    @Activate
    public LuxtronikHeatpumpHandlerFactory(final @Reference LocaleProvider localeProvider,
            final @Reference TranslationProvider i18nProvider) {
        this.translationProvider = new LuxtronikTranslationProvider(getBundleContext().getBundle(), i18nProvider,
                localeProvider);
    }
```

sgiehl replied:
the binding compiles for me as well without that line. But when replacing the built binding on my test server it throws the exception above when the new file is loaded. That doesn't happen when the line is present :thinking:

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java:
```diff
@@ -0,0 +1,1369 @@
+@NonNullByDefault
+public enum HeatpumpChannel {
+
+    // The channel ids are currently based on the variable names used in the heat pumps internal JAVA applet
+    // for possible values see https://www.loxwiki.eu/display/LOX/Java+Webinterface
+    // or https://github.com/Bouni/Home-Assistant-Luxtronik/blob/master/data.txt
```
It would be great if you normalize these and map them to English names.

sgiehl replied:
The channel names have been changed. I would like to still keep the names for the constants as they are, so it's still possible to map them to those of the internal java applet of the heatpump

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java:
```diff
@@ -0,0 +1,1369 @@
+        return requiredVisibility;
+    }
+
+    public Boolean isVisible(Integer[] visibilityValues) {
```
See above. Use primitives where possible. Please check all.
```suggestion
    public boolean isVisible(Integer[] visibilityValues) {
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java:
```diff
@@ -0,0 +1,1369 @@
+        HeatpumpVisibility visiblity = getVisibility();
+
+        if (visiblity == null) {
+            return Boolean.TRUE;
```
```suggestion
            return true;
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java:
```diff
@@ -0,0 +1,1369 @@
+
+        int code = visiblity.getCode();
+
+        if (visibilityValues.length < code || visibilityValues[code].equals(1)) {
```
```suggestion
        if (visibilityValues.length < code || visibilityValues[code] == 1) {
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpType.java:
```diff
@@ -0,0 +1,130 @@
+            }
+        }
+
+        logger.info("Unknown heatpump type code {}", code);
```
Logging to info should be used sparsely e.g. a newly started component or a user file that has been loaded. This could be debug or warn.

sgiehl replied:
I've choosen info so every user can see it by default. It only occurs if someone owns a heatpump that returns a type id that is not yet defined. Already added two additional types based on the feedback I got on the forum based on this log info. Will change it to warn as that should have the same effect I guess...

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,9 @@
+	xmlns:binding="https://openhab.org/schemas/binding/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>LuxtronikHeatpump Binding</name>
```
```suggestion
	<name>Luxtronik Heatpump Binding</name>
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump.properties:
```diff
@@ -0,0 +1,132 @@
+channel-type.luxtronikheatpump.HauptMenuStatus_Zeile3.state.option.13 = Heating ext. energy source
+channel-type.luxtronikheatpump.HauptMenuStatus_Zeile3.state.option.14 = Service water ext. energy source
+channel-type.luxtronikheatpump.HauptMenuStatus_Zeile3.state.option.16 = Flow monitoring
+channel-type.luxtronikheatpump.HauptMenuStatus_Zeile3.state.option.17 = Second heat generator 1 operation
+c
+
+channel-type.luxtronikheatpump.Ba_XX_akt.state.option.0 = Auto
```
Did you remove or add something accidentally?

sgiehl replied:
guess the `c` was added by accident...

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump.properties:
```diff
@@ -0,0 +1,132 @@
+channel-type.luxtronikheatpump.SwitchoffX_file_NrX.state.option.1 = Heat pump malfunction
+channel-type.luxtronikheatpump.SwitchoffX_file_NrX.state.option.2 = Facility malfunction
+channel-type.luxtronikheatpump.SwitchoffX_file_NrX.state.option.3 = Second heat generator operating mode
+channel-type.luxtronikheatpump.SwitchoffX_file_NrX.state.option.4 = EVU lock
```
Is EVU English?

sgiehl replied:
I guess `EVU` stands for `Energieversorgungsunternehmen`, not sure if there is a short equivalent in english. The heatpump uses `EVU` for english as well :thinking:

sgiehl replied:
Tried to find something with the same meaning on the internet. Seems "Utility lock" might work. So will use this for now

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+thing-type.config.luxtronikheatpump.heatpump.ipAddress.label = IP Adresse
+thing-type.config.luxtronikheatpump.heatpump.ipAddress.description = IP Adresse der Wärmepumpe
+thing-type.config.luxtronikheatpump.heatpump.port.label = Port
+thing-type.config.luxtronikheatpump.heatpump.port.description = Port der zum Aufbau einer Verbeindung zur Wärmepumpe verwendet werden soll. Standardwert ist 8889. Für Wärmepumpen mit einer Firmware version vor V1.73 muss der Port 8888 verwendet werden.
```
```suggestion
thing-type.config.luxtronikheatpump.heatpump.port.description = Port der zum Aufbau einer Verbindung zur Wärmepumpe verwendet werden soll. Standardwert ist 8889. Für Wärmepumpen mit einer Firmware version vor V1.73 muss der Port 8888 verwendet werden.
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.Temperatur_TRL.label = Rücklauftemperatur Heizkreis
+channel-type.luxtronikheatpump.Sollwert_TRL_HZ.label = Rücklauf-Soll Heizkreis
+channel-type.luxtronikheatpump.Temperatur_TRL_ext.label = Rücklauftemperatur im Trennspeicher.
+channel-type.luxtronikheatpump.Temperatur_THG.label = Heisgastemperatur
```
```suggestion
channel-type.luxtronikheatpump.Temperatur_THG.label = Heißgastemperatur
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.Sollwert_TRL_HZ.label = Rücklauf-Soll Heizkreis
+channel-type.luxtronikheatpump.Temperatur_TRL_ext.label = Rücklauftemperatur im Trennspeicher.
+channel-type.luxtronikheatpump.Temperatur_THG.label = Heisgastemperatur
+channel-type.luxtronikheatpump.Temperatur_TA.label = Aussentemperatur
```
Please check all.
```suggestion
channel-type.luxtronikheatpump.Temperatur_TA.label = Außentemperatur
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.Time_AbtIn.label = Zeit bis Abtauen
+channel-type.luxtronikheatpump.Temperatur_RFV2.label = Raumtemperatur Raumstation 2
+channel-type.luxtronikheatpump.Temperatur_RFV3.label = Raumtemperatur Raumstation 3
+channel-type.luxtronikheatpump.SH_SW.label = Schaltuhr Schwimmbad Symbol
```
```suggestion
channel-type.luxtronikheatpump.SH_SW.label = Schaltuhr Schwimmbadsymbol
```

sgiehl replied:
guess that would change the meaning. Maybe `Schaltuhr-Schwimmbad-Symbol` would be better?

fwolter replied:
Sounds good!

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.Temperatur_RFV3.label = Raumtemperatur Raumstation 3
+channel-type.luxtronikheatpump.SH_SW.label = Schaltuhr Schwimmbad Symbol
+channel-type.luxtronikheatpump.Zaehler_BetrZeitSW.label = Betriebsstunden Schwimmbad
+channel-type.luxtronikheatpump.FreigabKuehl.label = 	Freigabe Kühlung
```
```suggestion
channel-type.luxtronikheatpump.FreigabKuehl.label = Freigabe Kühlung
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.Zaehler_BetrZeitSW.label = Betriebsstunden Schwimmbad
+channel-type.luxtronikheatpump.FreigabKuehl.label = 	Freigabe Kühlung
+channel-type.luxtronikheatpump.AnalogIn.label = Analoges Eingangssignal
+# channel-type.luxtronikheatpump.SonderZeichen.label = SonderZeichen
+channel-type.luxtronikheatpump.SH_ZIP.label = Zirkulationspumpen Symbol
+# channel-type.luxtronikheatpump.WebsrvProgrammWerteBeobarten.label = ??????
```
Can the comments be removed?

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.FreigabKuehl.label = 	Freigabe Kühlung
+channel-type.luxtronikheatpump.AnalogIn.label = Analoges Eingangssignal
+# channel-type.luxtronikheatpump.SonderZeichen.label = SonderZeichen
+channel-type.luxtronikheatpump.SH_ZIP.label = Zirkulationspumpen Symbol
```
```suggestion
channel-type.luxtronikheatpump.SH_ZIP.label = Zirkulationspumpensymbol
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.LIN_VDH_out.label = Ausgang Verdichterheizung
+channel-type.luxtronikheatpump.HZIO_PWM.label = Steuersignal Umwälzpumpe
+channel-type.luxtronikheatpump.HZIO_VEN.label = Ventilator Drehzahl
+# channel-type.luxtronikheatpump.HZIO_EVU2.label = EVU2
+channel-type.luxtronikheatpump.HZIO_STB.label = Sicherheits-Tempeartur-Begrenzer Fussbodenheizung
+channel-type.luxtronikheatpump.SEC_Qh_Soll.label = Leistung Sollwert
+channel-type.luxtronikheatpump.SEC_Qh_Ist.label = Leistung Istwert
+channel-type.luxtronikheatpump.SEC_TVL_Soll.label = Temperatur Vorlauf Soll
+# channel-type.luxtronikheatpump.SEC_Software.label = Software Stand SEC Board
```
Can the comments be removed?

sgiehl replied:
Could remove them. I had decided to keep them, as they hold the descriptions that are used for those channels in the heatpumps java applet. not sure if those channels might make sense to use at some time....

fwolter replied:
I'd remove them, if they are not used, yet.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.LIN_ND.label = Niederdruck
+channel-type.luxtronikheatpump.LIN_VDH_out.label = Ausgang Verdichterheizung
+channel-type.luxtronikheatpump.HZIO_PWM.label = Steuersignal Umwälzpumpe
+channel-type.luxtronikheatpump.HZIO_VEN.label = Ventilator Drehzahl
```
```suggestion
channel-type.luxtronikheatpump.HZIO_VEN.label = Ventilatordrehzahl
```

sgiehl replied:
Guess that would be correct. Didn't look too much through the translations. I actually took them from the internal java applet. So it's what the heatpump itself uses. Will change it nevertheless...

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.SEC_FlussigEx.label = Flüssigtemperatur EEV (elektronisches Expansionsventil)
+channel-type.luxtronikheatpump.SEC_UK_EEV.label = Unterkühlung EEV
+channel-type.luxtronikheatpump.SEC_EVI_Druck.label = Druck EVI
+channel-type.luxtronikheatpump.SEC_U_Inv.label = 	Spannung Inverter
+channel-type.luxtronikheatpump.Temperatur_THG_2.label = 	Temperarturfühler Heissgas 2
```
There are some unintentional tabs. Please check all.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/i18n/luxtronikheatpump_de.properties:
```diff
@@ -0,0 +1,399 @@
+channel-type.luxtronikheatpump.Ba_XX_akt.state.option.4 = Aus
+
+channel-type.luxtronikheatpump.SwitchoffX_file_NrX.state.option.1 = Wärmepumpe Störung
+channel-type.luxtronikheatpump.SwitchoffX_file_NrX.state.option.2 = Anlagen Störung
```
```suggestion
channel-type.luxtronikheatpump.SwitchoffX_file_NrX.state.option.2 = Anlagenstörung
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="Temperatur_TVL">
+		<item-type>Number:Temperature</item-type>
+		<label>Flow temperature heating circuit</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars.

sgiehl replied:
I've capitalized them for now. might be hard to shorten some of the labels. Will need to think about them a bit and see if some better names will come to my mind...

sgiehl replied:
There are still a lot channel labels with more than 25 chars. But I'm not sure how to shorten them without changing the meaning. Is it ok to use abbreviations where possible?

fwolter replied:
I think it's ok to abbreviate temperature etc., but I wouldn't use abbreviations at any price.

fwolter reviewed (changes requested):
Thanks for this quite extensive binding! I reviewed your code and here is my feedback.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/ChannelUpdaterJob.java:
```diff
@@ -0,0 +1,274 @@
+        // setProperty("Type 2", heatpumpType2);
+
+        setProperty("Software version", getSoftwareVersion(heatpumpValues));
+        setProperty("IP address", transformIpAddress(heatpumpValues[91]));
```
in theory the configuration could also be done with a hostname that resolves to the ip. Can also remove all network related properties here or add them as channels instead. I don't have any preference :shrug:

fwolter replied:
I hadn't the hostname in mind. Adding them as Channels would be wrong, though. I leave it up to you if you keep or remove them.

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/HeatpumpConnector.java:
```diff
@@ -0,0 +1,208 @@
+     */
+    public void read() throws UnknownHostException, IOException {
+        try (Socket sock = new Socket(serverIp, serverPort)) {
+            InputStream in = sock.getInputStream(); // @SuppressWarnings
```
Iirc that was the same way in the old binding. didn't think much about it. will try to remove it

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/README.md:
```diff
@@ -0,0 +1,328 @@
+
+| channel  | type   | advanced | description                  |
+|----------|--------|----------|------------------------------|
+| Temperatur_TVL | Number:Temperature |   | Flow temperature heating circuit | 
```
I'm currently using the short names used in the java applet the heatpump uses. It's actually a lot work to change all of them again, also for some I technically don't know what they mean (based on the descriptions taken from the applet).
So could basically only use the current description as channel name, not sure if that would be helpful...

fwolter replied:
That would be better than the current state.

sgiehl replied:
ok. will try to do that the next days.

sgiehl replied:
I've renamed all channels now. Let me know if any channel name should still be renamed/improved

sgiehl commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/ChannelUpdaterJob.java:
```diff
@@ -0,0 +1,274 @@
+        }
+
+        returnValue += String.format("%02d:", Integer.valueOf(value / 3600));
+        value %= 3600;
```
ok. will change that.

sgiehl commented:
@fwolter thanks again for the valuable review & feedback. I hope I've addressed all comments now. 
Feel free to do another review and let me know if anything else needs to be adjusted :slightly_smiling_face:

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/README.md:
```diff
@@ -0,0 +1,331 @@
+
+## Channels
+
+As the Luxtronik 2 control is able to handle multiple heat pumps with different features (like heating, hot water, cooling, solar, photovoltaics, swimming pool,...), the binding has a lot channels. Depending on the heatpump it is used with, various channels might not hold any (useful) values. If `showAllChannels` is not activated for the thing, this binding will automatically try to hide channels ...
```
If you put each sentence in a separate line, reading the diffs would be easier.
```suggestion
As the Luxtronik 2 control is able to handle multiple heat pumps with different features (like heating, hot water, cooling, solar, photovoltaics, swimming pool,...), the binding has a lot channels. Depending on the heatpump it is used with, various channels might not hold any (useful) values. If `showAllChannels` is not activated for the thing, this binding will automatically try to hide channels that are not available for your heat pump. As this is done using reverse engineered parameters it might not be correct in all cases. If you miss a channel that should be available for your heat pump, you can enable `showAllChannels` for your thing, so all channels become available. Feel free to report such a case on the forum, so we can try to improve / fix that behavior.
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/README.md:
```diff
@@ -0,0 +1,331 @@
+| menuStateFull | String |   | Status (complete) |
+| menuStateLine1 | Number |   | Status line 1 |
+| menuStateLine2 | Number | x | Status line 2 |
+| menuStateLine3 | Number | x | Status Zeile 3 |
```
```suggestion
| menuStateLine3 | Number | x | Status line 3 |
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/README.md:
```diff
@@ -0,0 +1,331 @@
+| outputCompressorHeating | Switch | x | Output compressor heating |
+| controlSignalCirculatingPump | Number:Energy | x | Control signal circulating pump |
+| fanSpeed | Number | x | Fan speed |
+| temperatureSafetyLimitFloorHeating | Switch | x | Safety tempearture limiter floor heating |
```
```suggestion
| temperatureSafetyLimitFloorHeating | Switch | x | Safety temperature limiter floor heating |
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpType.java:
```diff
@@ -0,0 +1,131 @@
+            }
+        }
+
+        Logger logger = LoggerFactory.getLogger(HeatpumpType.class);
```
You can make the logger a static final field, if you need to access it in a static way, although if checkstyle is complaining, then.

sgiehl replied:
had it static in the first place, but changed that because of checkstyle. So it's fine to have a checkstyle complain in this case?

fwolter replied:
Yes.

sgiehl replied:
ok. is there any possibility to disable the check with a comment in the code or something like that?

fwolter replied:
Not that I'm aware of.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="temperatureHeatingCircuitFlow">
+		<item-type>Number:Temperature</item-type>
+		<label>Flow Temp. heating circuit</label>
```
```suggestion
		<label>Flow Temp. Heating Circuit</label>
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="temperatureHotGas" advanced="true">
+		<item-type>Number:Temperature</item-type>
+		<label>Hot Gas Temp.</label>
```
I think this looks a bit odd, if you abbreviate, although there is enough space.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="temperatureOutsideMean">
+		<item-type>Number:Temperature</item-type>
+		<label>Avg. Temp. Outside over 24 h (Heating Limit Func.)</label>
```
```suggestion
		<label>Avg. 24h-Temp. Outside (Limiting)</label>
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="inputASD" advanced="true">
+		<item-type>Switch</item-type>
+		<label>Input "Defrost End, Brine Pressure, Flow Rate"</label>
```
You could use the description tag for more text. Same for the inputs below.
```suggestion
		<label>Input Defrost/Brine/Flow"</label>
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="operationTimeSecondHeatGenerator2" advanced="true">
+		<item-type>Number:Time</item-type>
+		<label>Operation Time Second Heat Generator 2</label>
```
```suggestion
		<label>Runtime 2nd Heat Generator 2</label>
```

sgiehl replied:
I guess I might need to rename this or the other channel that actually has a similar name. Seems one contains the time it's currently running since and this one contains the total time it has run. guess i will use `runtimeCurrent` and `runtimeTotal` instead of operation time

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="heatingControllerMoreTime" advanced="true">
+		<item-type>Number:Time</item-type>
+		<label>Heating Controller More Ttime</label>
```
```suggestion
		<label>Heating Controller More Time</label>
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="errorTime0">
+		<item-type>DateTime</item-type>
+		<label>Timestamp Error 0 in Memory</label>
```
```suggestion
		<label>Timestamp Error 0</label>
```

sgiehl replied:
There are other channels for the values in `Memory 2`, but guess it's enough to mark those channels and remove the memory in this case :+1:

fwolter replied:
You could also group channels, if that helps.

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="timeLockSecondHotGasCompressor" advanced="true">
+		<item-type>Number:Time</item-type>
+		<label>Lock Second Compressor Hot Gas</label>
```
```suggestion
		<label>Lock 2nd Compressor Hot Gas</label>
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,2078 @@
+
+	<channel-type id="outputSupplyAirFan" advanced="true">
+		<item-type>Number:ElectricPotential</item-type>
+		<label>Supply Air Fan (Defrost Function)</label>
```
```suggestion
		<label>Supply Air Fan (Defrost)</label>
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,257 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<thing-type id="heatpump">
+		<label>Heatpump with Luxtronik control</label>
```
```suggestion
		<label>Heatpump with Luxtronik Control</label>
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,257 @@
+
+		<config-description>
+			<parameter name="ipAddress" type="text" required="true">
+				<label>IP address</label>
```
You could specify `<context>network-address</context>` to get a free validation.

sgiehl replied:
ah cool. will try that.

fwolter reviewed (changes requested):
Thanks for translating the channel names so quickly! They look pretty good! I made some suggestions how to shorten some labels further.

The code is already in a pretty good shape, too.

sgiehl commented:
@fwolter guess I have applied all suggested changes now

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,259 @@
+		<config-description>
+			<parameter name="ipAddress" type="text" required="true">
+				<context>network-address</context>
+				<label>IP address</label>
```
```suggestion
				<label>IP Address</label>
```

fwolter commented on bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,259 @@
+				<description>Port number of the heat pump to connect to. Defaults to 8889. For heatpumps running firmware version
+					before V1.73, port 8888 needs to be used.</description>
+			</parameter>
+			<parameter name="refresh" type="integer" required="false">
```
```suggestion
			<parameter name="refresh" type="integer" required="false" unit="s" min="1">
```

fwolter reviewed (changes requested):
Your changes look good!

fwolter reviewed (approved):
LGTM

As this is a new binding, another maintainer needs to review your code before it can be merged.

sgiehl commented:
@fwolter FYI: due to some feedback on the forum I had to push another small commit to fix some incorrectly handled channels

fwolter reviewed (approved):
Due to lack of review capacity, the rules for new bindings were weakened. New bindings need only the approval of one reviewer instead of two from now on.

sgiehl commented:
thanks for merging and your support @fwolter
